src/fastpath/fastpath.c

```diff
@@ -6,15 +6,175 @@
 
 #include <config.h>
 #include <fastpath/fastpath.h>
-#ifdef CONFIG_KERNEL_MCS
-#include <object/reply.h>
-#endif
 
 #ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
 #include <benchmark/benchmark_track.h>
 #endif
 #include <benchmark/benchmark_utilisation.h>
 
+#ifdef CONFIG_SIGNAL_FASTPATH
+#ifdef CONFIG_ARCH_ARM
+static inline
+FORCE_INLINE
+#endif
+void NORETURN fastpath_signal(word_t cptr, word_t msgInfo)
+{
+    word_t fault_type;
+    sched_context_t *sc = NULL;
+
+    /* Get fault type. */
+    fault_type = seL4_Fault_get_seL4_FaultType(NODE_STATE(ksCurThread)->tcbFault);
+
+    /* Check there's no saved fault */
+    if (unlikely(fault_type != seL4_Fault_NullFault)) {
```

Comment:
@lsf37, is this necessary for proof reasons? Isn't there an invariant that a ksCurThread's FaultType is always seL4_Fault_NullFault if it's invoking the fastpath?

Reply:
I don't think we have proved that invariant. It sounds something like `thread_state t = Running --> fault t = None` might be true in general, but I couldn't find anything in the current proofs right now. @Xaphiosis or @corlewis might remember more.

I'd vote for leaving this in, but adding a comment that this could be removed if we can prove that the current thread can't have a fault when it is invoking the fastpath.

Reply:
Apologies for the extremely late reply, but an invariant equivalent to this has been proven to hold for the MCS abstract specification.

Reply:
Ok, should I remove it now or leave it for now to maintain consistency with the seL4_Call fastpath so that it can be removed when the check is removed from that?